### PR TITLE
Use specialized constructors and add scoping

### PR DESCRIPTION
**Pull request by k0ekk0ek**
This PR uses specialized constructors for non-terminals in the parser which makes it much easier to implement semantic checks because memory management becomes a lot easier (less branches) and keeps the grammar readable. It also improves scoping. It's not perfect yet, inheritance of struct types still needs to be added, but it's good enough to ensure the parent pointer in tree nodes is set properly and should provide a decent base for future work.

### CODE REVIEW

**Comment by k0ekk0ek** [comment]
I should note that I removed `#pragma` support for the time being. It isn't that much work to re-add it, most of the code can be reused, but due to time constraints (and it isn't on the list for things that we must support in the initial version) I chose to save myself the hassle.

**Comment by k0ekk0ek** [comment]
@eboasson, I know there's two macOS builds failing, but that's due to Maven. (MacStadium probably has some servers that aren't whitelisted with the Maven repository servers which is causing us to run into the maximum number of downloads being exceeded or something). With your approval I'd like to hit merge on this. It's certainly not good enough to go into master, but there's some refactoring/rework that I have planned after we get the initial version done anyway...

**Review by eboasson — APPROVED** [review]
Go ahead @k0ekk0ek and merge it. The two comments I added do not require any changes at this point in time, and the macOS builds have been running into this infrastructure problem for a while now, and short of moving to another CI system, it seems like there is little we can do about it right now.

File: src/idl/include/idl/tree.h [hunk 1]
```diff
@@ -234,68 +247,50 @@ typedef struct idl_sequence idl_sequence_t;
 struct idl_sequence {
   idl_node_t node;
   idl_simple_type_spec_t *type_spec;
-  idl_const_expr_t *const_expr;
+  uint64_t maximum; /**< maximum size of bounded sequence or 0 if unbounded */
```
[the remark this answers is not in the document]
Comment (k0ekk0ek) [reply src/idl/include/idl/tree.h:250]: Annotated in my WiP branch

File: src/idl/include/idl/tree.h [hunk 2]
```diff
@@ -212,11 +220,16 @@ typedef struct idl_constval idl_constval_t;
...
   union {
-    uint8_t oct;
-    int32_t lng;
-    uint32_t ulng;
-    int64_t llng;
-    uint64_t ullng;
+    bool bln;
+    char chr;
+    int8_t int8;
+    uint8_t uint8, oct;
```
[the remark this answers is not in the document]
Comment (k0ekk0ek) [reply src/idl/include/idl/tree.h:226]: I need to treat e.g. int32 and long differently in the future for better error messages even though they're the same size, but you'll get a chance to be strict soon (hopefully) :sweat_smile:

**Comment by k0ekk0ek** [comment]
Thanks @eboasson

**Comment by k0ekk0ek** [comment]
Btw, I should not that switching CI providers won't do much good. I think most use MacStadium for macOS builds. Not 100% sure though. I'll open a ticket with Travis

**Pull request merged**